src/runtime_config.rs

```diff
@@ -0,0 +1,94 @@
+use serde::Deserialize;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::fs;
+use tokio::sync::RwLock;
+use tokio::task::JoinHandle;
+use tracing::{error, info};
+
+#[derive(Debug, Deserialize, Clone, PartialEq, Default)]
+pub struct RuntimeConfig {
+    /// A list of tasks to drop before inserting into sqlite.
+    pub drop_task_killswitch: Vec<String>,
+}
+
+pub struct RuntimeConfigManager {
+    pub config: Arc<RwLock<RuntimeConfig>>,
+    pub handler: JoinHandle<()>,
+}
+
+impl RuntimeConfigManager {
+    pub async fn new(path: Option<String>) -> Self {
+        let runtime_config = Arc::new(RwLock::new(Default::default()));
+        let _ = Self::reload_config(&path, &runtime_config).await;
+        Self {
+            config: runtime_config.clone(),
+            handler: tokio::spawn(async move {
+                let mut interval = tokio::time::interval(Duration::from_secs(60));
+                loop {
+                    Self::reload_config(&path, &runtime_config).await;
+                    interval.tick().await;
+                }
+            }),
+        }
+    }
+
+    async fn reload_config(path: &Option<String>, config: &Arc<RwLock<RuntimeConfig>>) {
+        if path.is_none() {
+            return;
+        }
+        let yaml_path = path.as_ref().unwrap();
+        let contents = fs::read_to_string(yaml_path).await;
+        let new_config: RuntimeConfig = match contents {
+            Ok(contents) => {
+                let runtime_config = serde_yaml::from_str::<RuntimeConfig>(&contents);
+                match runtime_config {
+                    Ok(runtime_config) => runtime_config,
+                    Err(e) => {
+                        error!("Using default runtime configs. Failed to parse file: {}", e);
+                        RuntimeConfig::default()
+                    }
+                }
+            }
+            Err(e) => {
+                error!(
+                    "Using default runtime configs. Failed to read yaml file: {}",
+                    e
+                );
+                RuntimeConfig::default()
+            }
+        };
+        if new_config != *config.read().await {
+            *config.write().await = new_config;
+            info!("Reloaded new runtime config from {}", yaml_path);
+        }
+    }
+
+    pub async fn read(&self) -> RuntimeConfig {
+        let m = self.config.read().await;
+        m.clone()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::RuntimeConfigManager;
+    use tokio::fs;
+
+    #[tokio::test]
+    async fn test_runtime_config_manager() {
+        let test_yaml = r#"
+drop_task_killswitch:
```

Comment:
It would be good to have additional tests covering

- The config file missing.
- The config file containing invalid yaml.
 
As Evan mentioned we should preserve the existing configuration in both of these scenarios as we need to be resilient to bad configuration deploys.

Reply:
Addressed Evan's point and added 3 other tests: missing config test, invalid yaml test, and preserving existing configs test if reload sees either a missing or invalid yaml.